### ISSUE TITLE
Propagate extrinsic errors over JSON-RPC for locally submitted transactions

### ISSUE THREAD

**Issue by foby** [paritytech/substrate#2246]
Extrinsic errors raised within module functions are not propagated over the JSON-RPC interface. The executor indeed handles those errors by sending an `ExtrinsicFailed` event. But that event doesn't carry any further information about the error.

Example:

In our module transaction we raise this error: 
```
return Err("my_error_code")
```

And on the client side the transaction result is like:
```
{
  "events": [
    {
      "phase": {
        "ApplyExtrinsic": 1
      },
      "event": {
        "index": "0x0001",
        "data": []
      }
    }
  ],
  "status": {
    "Finalised": "0xf813791343ba383cb28e4a4b2df2e2948bff1c973de93c1ce404d5d78c335d70"
  },
  "type": "Finalised"
}
```

So we get the `ExtrinsicFailed` event (0x0001) but with an empty `data` field. 

Suggestion:
Is it possible to encode the error object into the `data` field of the event record?

**Comment by xlc** [comment]
We also really wants this feature because otherwise it is very hard for client side to figure out what's going on.

I can implement this if that's ok.  Just let me know if there are anything I need to be caution about.

**Comment by bkchr** [comment]
CC @tomusdrw

**Comment by tomusdrw** [comment]
It's related more to how `srml/system` notes the applied extrinsics than the extrinsic itself, I'm not sure though if it's a good idea to store arbitrary (unbounded) strings inside the event, it seems it could be heavily abused.
We could maybe store few first byte of the hash of the (trimmed) message or better just error codes.

**Comment by xlc** [comment]
I think the easy way is trunk the error message if it is more than X chars (30?).
Hash will work but it is hard to consume on SDK side and it is not human readable.
Error codes will be the proper way but I don't know how to do it without change the `dispatch::Result` type and refactor every modules.

**Comment by tomusdrw** [comment]
> I think the easy way is trunk the error message if it is more than X chars (30?).
Sure, but it's really wasteful cause we are talking about 30 bytes vs 1 byte for every failed transaction in the chain, hence why I thought on some other ideas. Maybe we could truncate to a single byte actually :)

**Comment by foby** [comment]
I'm in favor of numeric error codes (`u16` should be large enough). Processing numbers is straightforward and less error prone. Translation into something human readable can be made on the client side.

**Comment by xlc** [comment]
Ok. So everyone wants `u16` or `u8` error code. Then it should really be an enum.
Now the question is, how to do it? Currently the errors are `&'static str`. Should I change `support::dispatch::Result` to `Result<(), u8>` and refactor every modules to convert static string to enum?
And this needs to be in 1.0 right? as it is a breaking change.

**Comment by xlc** [comment]
I am thinking have `decl_module` defines a type `DispatchError` which bounded to `Into<u16>`

and also macro defines/rewrite `Result` to be `Result<(), DispatchError>`.

Not sure if something like `Result<(), impl Into<DispatchError>>` will work.
Then make `DispatchError` have a variant `Unknown(&'static str)` which encode to `0`, that basically means unknown error.

So existing code will have error code of 0 without rewrite, not good enough but not a breaking change. New code can define new errors and return a proper error code.

Let me know if this sounds feasible.

**Comment by bkchr** [comment]
Sounds like a plan, but how do we make error codes unique across modules?

Okay I think got some sort of idea:

```
enum ModuleError {
     Unknow(&'static str),
     MyError0
}
```

First we need to expose the error in the metadata as well. Then, when we encounter some error, we calculate the error code in the following way:
1. Module index * 100 (I think 100 errors per module are more than enough).
2. Error as u16 + index from 1 = error code

Thus we have unique indices + a way to get an error description (via metadata).

**Comment by bkchr** [comment]
If the user does not specifies any error, we would just create the following error:
```
enum ModuleError {
     Unknow(&'static str),
}
```

**Comment by xlc** [comment]
I don't think we have to make error code unique because we know which tx cause the error hence we know which module the error belongs to.

That just means a module cannot return the error code from another module directly.
Also shared error type may need some thoughts like [`Currency::transfer`](https://github.com/paritytech/substrate/blob/1faed993e6b6ced8071e5bfe903ea6c1dd5d6c44/srml/support/src/traits.rs#L293)? have a `CurrencyError` enum? but the `Currency` implementing module won't be able to return custom error code to indicate some custom error. I guess we can make the `ErrorType` generic in `Currency`, not sure what is the best practice of Rust error propagation.

**Comment by bkchr** [comment]
Hmm yeah, I did not thought about cross module errors..

**Comment by xlc** [comment]
A possible (but complicated) solution is like how the `Call` or `Event` is defined for runtime. We can have something similar for the `RuntimeError` type. 

Then the macro should generate something like

```
enum RuntimeError {
  SystemError(system::ModuleError),
  BalanceError(balances::ModuleError),
  StakingError(staking::ModuleError),
}
```

and implements `From<system::ModuleError>`, `From<balances::ModuleError>`, `From<staking::ModuleError>` etc.

Because we are using `u16` or two bytes, we can have first byte as the module index, and the second byte as the module error index. So all the `ModuleError` needs to `Into<u8>` and we can't have more than 256 modules in a runtime.

**Comment by bkchr** [comment]
Yeah, that was part of my plan, having the `construct_runtime!` generate the `RuntimeError`.

**Comment by xlc** [comment]
@bkchr I am planning implement this feature but have few questions.

My plan is:

Modify `ApplyResult` by make it a two byte value (same as the current one).

First byte: module index (same as metadata module index)
Second byte: module specific error value.

Each module will have its `Error` enum implements `Into<u8>`. The system one will have a special value `system::Error::success` which is `0`. This means `0x0000` ApplyResult means success (same as current).

There will be a `From<'static str>` for `Error` and results some `Unknown` error for source code compatibly reason.

The question is I am not sure if this is going to break compatibility as `ApplyError` is used by `BlockBuilder` runtime api.

https://github.com/paritytech/substrate/blob/b78dc0024439cbf4d5ce4f6cbd937eb7707f6831/core/client/src/block_builder/api.rs#L29

Do I need to bump `#[api_version(3)]`? Or just add a new `apply_extrinsic` and decorate it with `#[changed_in(3)]` and keep the old one for backward compatibility?

The binary interface renames the same (ApplyResult is still two byte value) and the happy path value renames the same (0x0000 is still success).

**Comment by bkchr** [comment]
You will need to bump the `api_version` and would need to decode differently based on the version. 
Why do you want to handle `system` special? Just for backwards compatibility?

**Comment by tomusdrw** [comment]
I believe this was addressed by #3433. Shall we close?

**Comment by bkchr** [comment]
Yeah :+1:
